File: e-scala/src/main/scala/e/scala/codec/Codec.scala
package e.scala.codec

import e.scala.EOr

import scala.annotation.implicitNotFound

/**
 * Typeclass defining decoding and encoding together via [[e.scala.codec.Decoder]] and [[e.scala.codec.Encoder]]
 *
 * @tparam S Type of source
 * @tparam T Type of target
 */
@implicitNotFound(
  "No implicit instance is found of type e.scala.codec.Codec[${S}, ${T}]. You may try following:\n\n" +
  "* Make sure an instance of correct types is in scope (missing import?)\n" +
  "* Implement an implicit instance yourself\n" +
  "* Ensure both e.scala.codec.Decoder[${T}, ${S}] and e.scala.codec.Encoder[${S}, ${T}] are in scope and use `Codec.of[${S}, ${T}]`"
)
trait Codec[S, T] extends Decoder[T, S] with Encoder[S, T]

object Codec {
  /**
   * Summons an implicit codec
   *
   * @param codec Implicit instance of codec
   *
   * @tparam S Type of source
   * @tparam T Type of target
   *
   * @return Provided implicit codec
   */
  def apply[S, T](implicit codec: Codec[S, T]): Codec[S, T] = codec

  /**
   * Creates a codec based on an implicit decoder and encoder
   *
   * @param decoder Implicit instance of decoder
   * @param encoder Implicit instance of encoder
   *
   * @tparam S Type of source
   * @tparam T Type of target
   *
   * @return Created codec
   */
  def of[S, T](implicit decoder: Decoder[T, S], encoder: Encoder[S, T]): Codec[S, T] =
    new Codec[S, T] {
      override def decode(input: T): EOr[S] = decoder.decode(input)
      override def encode(input: S): T      = encoder.encode(input)
    }
}

File: e-java/src/main/java/e/java/codec/Codec.java
package e.java.codec;

import e.java.EOr;

/**
 * Typeclass defining decoding and encoding together via {@link e.java.codec.Decoder} and {@link e.java.codec.Encoder}
 *
 * @param <S> Type of source
 * @param <T> Type of target
 */
public interface Codec<S, T> extends Decoder<T, S>, Encoder<S, T> {
    /**
     * Creates a codec based on an implicit decoder and encoder
     *
     * @param <S> Type of source
     * @param <T> Type of target
     *
     * @param decoder Implicit instance of decoder
     * @param encoder Implicit instance of encoder
     *
     * @return Created codec
     */
    static <S, T> Codec<S, T> from(Decoder<T, S> decoder, Encoder<S, T> encoder) {
        return new Codec<S, T>() {
            @Override public EOr<S> decode(T input) {
                return decoder.decode(input);
            }

            @Override public T encode(S input) {
                return encoder.encode(input);
            }
        };
    }
}
